File: cracking_the_coding_interview/3-4.cpp
#include <iostream>
#include <stack>
using namespace std;

template <typename T>
class MyQueue {
public:
    MyQueue(){}

    void push(T a){
        while(!queueStack.empty()){
            preStack.push(queueStack.top());
            queueStack.pop();
        }
        queueStack.push(a);
        while(!preStack.empty()){
            queueStack.push(preStack.top());
            preStack.pop();
        }
    }

    void pop(){
        queueStack.pop();
    }

    T top(){
        return queueStack.top();
    }

    bool empty(){
        return queueStack.empty();
    }

private:
    stack<T> preStack;
    stack<T> queueStack;
};

int main(void){
    MyQueue<int> myQueue;
    myQueue.push(0);
    myQueue.push(1);
    myQueue.push(2);
    myQueue.pop();
    myQueue.push(3);
    while(!myQueue.empty()){
        cout << myQueue.top() << " ";
        myQueue.pop();
    }
    cout << endl;
    return 0;
}

File: cracking_the_coding_interview/3-4.py
class MyQueue(object):
    def __init__(self):
        self.s1 = []
        self.s2 = []

    def put(self, val):
        while len(self.s2) > 0:
            self.s1.append(self.s2.pop())
        self.s1.append(val)
        return
    
    def get(self):
        if not self.s1 and not self.s2:
            return None
        while len(self.s1) > 1:
            self.s2.append(self.s1.pop())
        return self.s1.pop() if len(self.s1) > 0 else self.s2.pop()

if __name__ == "__main__":
    q = MyQueue()
    q.put(0)
    q.put(1)
    q.put(2)
    print(q.get()) # 0
    q.put(3)
    print(q.get()) # 1
    print(q.get()) # 2
    print(q.get()) # 3
    print(q.get()) # None
    q.put(4)
    print(q.get()) # 4